crates/blockchain/blockchain.rs

```diff
@@ -116,17 +108,12 @@ pub fn find_parent_header(
     block_header: &BlockHeader,
     storage: &Store,
 ) -> Result<BlockHeader, ChainError> {
-    let parent_hash = block_header.parent_hash;
-    let parent_number = storage.get_block_number(parent_hash)?;
-
-    if let Some(parent_number) = parent_number {
-        let parent_header = storage.get_block_header(parent_number)?;
+    let Some(parent_header) = storage.get_block_header_by_hash(block_header.parent_hash)? else {
+        return Err(ChainError::ParentNotFound);
+    };
 
-        if let Some(parent_header) = parent_header {
-            Ok(parent_header)
-        } else {
-            Err(ChainError::ParentNotFound)
-        }
+    if parent_header.number == block_header.number - 1 {
+        Ok(parent_header)
     } else {
         Err(ChainError::ParentNotFound)
```

Comment:
nit: here we have a slightly different issue, which is that the block number of the new block is not correct.

Reply:
I was going to change this to "InvalidHeader" but I checked and this validation was already present in the `validate_header` validation, which makes sense. Removed the check here altogether.